xarray/core/rolling.py

```diff
@@ -203,26 +214,18 @@ def reduce(self, func, **kwargs):
             Array with summarized data.
         """
 
-        windows = [window.reduce(func, dim=self.dim, **kwargs)
-                   for _, window in self]
-
-        # Find valid windows based on count
-        if self.dim in self.obj.coords:
-            concat_dim = self.window_labels
-        else:
-            concat_dim = self.dim
-        counts = concat([window.count(dim=self.dim) for _, window in self],
-                        dim=concat_dim)
-        result = concat(windows, dim=concat_dim)
-        # restore dim order
-        result = result.transpose(*self.obj.dims)
+        windows = self.to_dataarray('_rolling_window_dim')
+        result = windows.reduce(func, dim='_rolling_window_dim', **kwargs)
 
+        # Find valid windows based on count.
+        # We do not use `reduced.count()` because it constructs a larger array
+        # (notice that `windows` is just a view)
+        counts = (~self.obj.isnull()).rolling(
```

Comment:
Maybe we should add a short-cut here that doesn't bother to compute `counts` if the array's dtype cannot hold NaN? I think that would solve the issue with changing `maybe_promote` for booleans.

You could add a utility function to determine this based on whether the result of `maybe_promote()` has the same dtype as the input.